Small steps

gasche opened the pull request:
This was fun.

o- commented on rewrite.ml:
```diff
@@ -0,0 +1,182 @@
+open Instr
+
+(* TODO remove the fallthrough cases *)
```
we did that

o- replied:
@gasche you have to fix it, the PR is on your fork :)

gasche replied:
Sorry, fixed in fdf6cc61481d017cdc6502fc0e540a9e6ec28c12.

mhyee commented on edit.ml:
```diff
@@ -0,0 +1,46 @@
+
+type result = Instr.instruction_stream * pc_map
+
+let subst_pc instrs start_pc len new_instrs pc =
```
Could you add some documentation for this (and also `subst` and `subst_many`)? For example, what the parameters mean, and what the function is actually used for.

For `subst`, it didn't occur to me that `len` could be `0`, until I saw it being used that way. Probably because I thought `len` referred to the length of `new_instrs`.

For `subst_many`, documenting the `substs` parameter would have been helpful. I also object to reusing the variable name `rest` for completely different things (i.e. the rest of `substs` and the rest of the result).

mhyee commented on instr.ml:
```diff
@@ -261,3 +261,8 @@ module Value = struct
 end
+
+let instr_at instrs pc =
+  if Array.length instrs = pc then Stop
```
What if `Array.length instrs > pc`? Did I miss something?

mhyee commented on rewrite.ml:
```diff
@@ -0,0 +1,196 @@
+open Instr
+
+(** [split_edge instrs pc label pc']
```
This documentation is good, but it's still not obvious what "splitting the edge" means. I think a short statement like "The result is that [pc] is no longer a predecessor of [pc']." Or maybe even a short example.

mhyee commented on rewrite.ml:
```diff
@@ -0,0 +1,196 @@
+
+type progress = Made_progress | Stuck
+
+let debug =false
```
Is this still used?

mhyee commented on tests.ml:
```diff
@@ -571,6 +570,290 @@ let do_test_codemotion = function () ->
+    let instrs = match Rewrite.pull_drop input x pc with
+      | Pulled_to ((instrs, _), _) -> instrs
+      | Blocked -> input in
+    (* Printf.printf "%s:\n%s\n"
```
Is this printing for debugging? Or should it just be removed (instead of being commented out)?

There's another instance of this below, for the push drop tests.

mhyee commented on tests.ml:
```diff
@@ -663,6 +946,8 @@ let suite =
    "used">:: do_test_used;
    "liveness">:: do_test_liveness;
    "codemotion">:: do_test_codemotion;
+   (* "push_drop">:: do_test_push_drop; *)
```
Why is this commented out?

mhyee reviewed (commented):
I'm going to be that person who asks for more documentation. I've commented on some of the places I had trouble following.

More generally, I think the whole `push_drop` and `pull_drop` needs some sort of description or summary. Same with `move_drops` and `move_drops_on_var`. The email from a while back was good, but it's something I had to dig up when reviewing this code. For example, I liked the summary:

>   - if the instruction depends on `x`, you cannot move above
>  
>   - if the instruction does not use `x` at all, you can move
>     above
> 
>   - if the instruction assigns `x` (`x <- e`), you can move
>     above *and* delete the instruction (optimization!);
>     (for `read` we would need an anonymous read `read _` to
>     leave in place)
> 
>   - if the instruction declares `x`, you can delete it *and*
>     delete the drop as well.

Explaining the need for edge splitting (and defining what it exactly is) would be good.

Also, just from reading the names "push" and "pull," it's not clear what "direction" they're referring to. I can't think of any short names that would make this clearer, so this is just another request for documentation.

Finally, it seems there's a few merge conflicts. Maybe there's some GitHub issue -- it looked like the commits were out of order. I'm not sure what happened.

gasche commented:
These are excellent comments, thanks. I'll try to iterate to fix those as soon as possible.

o- commented:
@gasche @mhyee  
I am currently working on a PR that abstracts over the drop moving to have a generic move-instruction version. If you all agree I would suggest that we merge this PR and I can address the review in the new PR? Because otherwise it might become a bit messy.

gasche commented:
I will do one round of iteration tomorrow morning to fix the conflicts and merge the branch as you suggest. I may fix some of the issues reported at the same time.

mhyee commented:
Yeah, that sounds good.

o- committed:
factor out the move-one-instruction algorithm from move_drop

make it generic so we can reuse the same implementation for other
code motions.

o- commented:
@gasche  
oh, ok. I was too fast on this one, sorry.
https://github.com/reactorlabs/sourir/commit/df7ddc33d0896053ce5f072494ad217710190c05
What shall we do? If you want to clean up this one first you can go ahead. I'll figure out how to rebase. Otherwise you can cherry-pick that commit and continue on top in this PR?

gasche commented:
I pushed your commit on top of the PR. If you make other commits to your
`move-it` branch I can push them as well.

P.S.: We are starting to have a bit too many branch in our forks; I did a
round of `git branch -d ...`, and then `git push --all --prune gasche`.

On Thu, Mar 16, 2017 at 9:36 PM, its me <notifications@github.com> wrote:

> oh, ok. I was too fast on this one, sorry.
> df7ddc3
> <https://github.com/reactorlabs/sourir/commit/df7ddc33d0896053ce5f072494ad217710190c05>
> What shall we do? If you want to clean up this one first you can go ahead.
> I'll figure out how to rebase. Otherwise you can cherry-pick that commit
> and continue on top in this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/reactorlabs/sourir/pull/65#issuecomment-287183425>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAaA_mB68I4OOy-Dwv3P5csD7hPKZpjqks5rmZ1ggaJpZM4MaaDQ>
> .
>

o- commented:
@gasche 
thank you! I might start playing with moving around other instructions. For example assignments or osr.

mhyee commented on rewrite.ml:
```diff
@@ -0,0 +1,228 @@
+  | Work of Edit.result * pc list
+
+let push_instr is_eliminating is_anihilating is_blocking instrs pc : push_status =
+  assert (pc >0);
```
Nitpick: missing space.

mhyee commented on rewrite.ml:
```diff
@@ -0,0 +1,228 @@
+    let move_up_instr = move_up_instr is_eliminating is_anihilating is_blocking in
+
+    (* Loop around and move all drops until none moves anymore *)
+    let rec work status instrs pc =
```
This function is a bit confusing, because `instrs` is reused. It's particularly bad in lines 171/172 where both cases return `instrs` but they're different. 177/178 is similar.

mhyee replied:
It's actually worse because there are *three* `instrs`. One is an argument, one is the let binding on line 170, and one is from `Some instrs`.

mhyee commented on rewrite.ml:
```diff
@@ -0,0 +1,228 @@
+        work status instrs (pc+1)
+    in
+    work Stuck instrs 0
+
```
Delete this empty line?

mhyee commented on rewrite.ml:
```diff
@@ -0,0 +1,228 @@
+
+
+module Drop = struct
+  let is_blocking var instr =
```
Note that in the arguments of `move_up` and on lines 94-96, the order is `is_eliminating`, `is_annihilating`, (it's spelled with two **n**s) and `is_blocking`. But here, the order is `is_blocking`, `is_eliminating`, and `is_annihilating`.

mhyee commented on rewrite.ml:
```diff
@@ -0,0 +1,228 @@
+      | Drop x when x = var -> true
+      | _ -> false
+    in
+    let move = move_up
```
It's a bit confusing to have two different functions called `move_up`. Especially when it's **not** the function defined just a few lines later.

mhyee commented:
I added some more comments on code style. Other than that, it looks fine.

Generalizing the algorithm for other instructions is good, but I wonder if it's general enough (but not too general). It'll probably be fine, and we can tweak it later. I'm basically citing [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) and the [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)), but since we do want to move other instructions, I think the refactor is justified.

o- commented:
@mhyee thanks a lot for looking at it. My patch was a bit sloppy as I was rushing to get it done to not block progress. It does not even address your earlier comments. Don't waste too much time on it. I am not sure yet if it will be general enough/too general. But I think we'll see. At least the moves I am playing around with right now seem to mostly use is_blocking. especially is_annihilating is very particular to drop. But the whole machinery to drive the moves is certainly a general thing. Also there seems to be some bugs regarding termination. At least my osr moving loops forever. I guess I'll have another patch soon :)

gasche commented:
I just rebased, and plan to merge.

gasche committed:
take Ming-Ho's remarks into account

gasche committed:
Rewrite: rename pull_instr into try_pull and move_up into pull

gasche commented:
So I merged the first part (the part Oli and I did together, for which @mhyee's comments had already been handled), and I applied some additional patches suggested by @mhyee's comments on the last commit by @o-. Any final comments?

gasche committed:
capture the eliminating/annihilating/blocking functions in a record

o- commented:
lgtm, thanks for the additional cleanup.
